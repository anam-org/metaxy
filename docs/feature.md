## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
