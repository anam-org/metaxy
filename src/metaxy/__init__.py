from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
