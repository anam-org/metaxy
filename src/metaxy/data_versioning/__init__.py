"""Data versioning module for sample-level data version calculation."""

from metaxy.data_versioning.calculators import (
    DataVersionCalculator,
    PolarsDataVersionCalculator,
)
from metaxy.data_versioning.diff import (
    DiffResult,
    MetadataDiffResolver,
    PolarsDiffResolver,
)
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.data_versioning.joiners import PolarsJoiner, UpstreamJoiner

__all__ = [
    "HashAlgorithm",
    "UpstreamJoiner",
    "PolarsJoiner",
    "DataVersionCalculator",
    "PolarsDataVersionCalculator",
    "DiffResult",
    "MetadataDiffResolver",
    "PolarsDiffResolver",
]
