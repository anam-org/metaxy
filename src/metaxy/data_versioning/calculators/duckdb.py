"""DuckDB-specific data version calculator with extension management.

This calculator extends IbisDataVersionCalculator to handle DuckDB-specific
extension loading (e.g., hashfuncs for xxHash support).
"""
# pyright: reportImportCycles=false

from collections.abc import Mapping, Sequence
from typing import TYPE_CHECKING, Any

from metaxy.data_versioning.calculators.ibis import IbisDataVersionCalculator
from metaxy.data_versioning.hash_algorithms import HashAlgorithm

if TYPE_CHECKING:
    import ibis

    from metaxy.data_versioning.calculators.ibis import HashSQLGenerator
    from metaxy.metadata_store.duckdb import ExtensionInput


class DuckDBDataVersionCalculator(IbisDataVersionCalculator):
    """DuckDB-specific calculator that manages extensions lazily.

    This calculator:
    1. Installs and loads DuckDB extensions on first use (lazy loading)
    2. Supports xxHash64, xxHash32, and MD5 hash functions
    3. Generates DuckDB-specific SQL for hash computation

    The extension loading happens in __init__, which is only called when
    native data version calculations are actually needed (not on store open).

    Example:
        >>> backend = ibis.duckdb.connect("metadata.db")
        >>> calculator = DuckDBDataVersionCalculator(
        ...     backend=backend,
        ...     extensions=["hashfuncs"]
        ... )
        >>> # Extensions are now loaded and xxHash64 is available
    """

    def __init__(
        self,
        backend: "ibis.BaseBackend",
        extensions: "Sequence[ExtensionInput] | None" = None,
    ):
        """Initialize DuckDB calculator and load extensions.

        Args:
            backend: DuckDB Ibis backend connection
            extensions: List of DuckDB extensions to install/load.
                Can be strings (from 'community' repo) or dicts with
                'name' and optional 'repository' keys.

        Example:
            >>> extensions = ["hashfuncs"]  # Simple form
            >>> extensions = [{"name": "spatial", "repository": "core_nightly"}]
        """
        self._backend = backend
        self.extensions = list(extensions or [])

        # Load extensions immediately (lazy at calculator creation time)
        self._load_extensions()

        # Generate hash SQL generators for DuckDB
        hash_sql_generators = self._generate_hash_sql_generators()

        # Initialize parent with backend and generators
        super().__init__(
            backend=backend,
            hash_sql_generators=hash_sql_generators,
        )

    def _load_extensions(self) -> None:
        """Install and load DuckDB extensions.

        This is called once when the calculator is created, which happens
        lazily when native data version calculations are first needed.
        """
        if not self.extensions:
            return

        # Type narrowing: we know this is a DuckDB backend
        from typing import cast

        backend = cast(
            Any, self._backend
        )  # DuckDB backend has raw_sql but not in ibis.BaseBackend stubs

        for ext_spec in self.extensions:
            if isinstance(ext_spec, str):
                ext_name = ext_spec
                # Install and load extension from community
                backend.raw_sql(f"INSTALL {ext_name} FROM community")
                backend.raw_sql(f"LOAD {ext_name}")
            else:
                ext_name = str(getattr(ext_spec, "name", ""))
                ext_repo = getattr(ext_spec, "repository", None) or "community"

                if ext_repo == "community":
                    # Install from community repository
                    backend.raw_sql(f"INSTALL {ext_name} FROM community")
                else:
                    # Set custom repository and install
                    backend.raw_sql(f"SET custom_extension_repository='{ext_repo}'")
                    backend.raw_sql(f"INSTALL {ext_name}")

                # Load extension
                backend.raw_sql(f"LOAD {ext_name}")

    def _generate_hash_sql_generators(self) -> dict[HashAlgorithm, "HashSQLGenerator"]:
        """Generate hash SQL generators for DuckDB.

        DuckDB supports:
        - MD5: Always available (built-in)
        - XXHASH32, XXHASH64: Available when 'hashfuncs' extension is loaded

        Returns:
            Dictionary mapping HashAlgorithm to SQL generator functions
        """
        generators: dict[HashAlgorithm, HashSQLGenerator] = {}

        # MD5 is always available
        def md5_generator(table, concat_columns: dict[str, str]) -> str:
            hash_selects: list[str] = []
            for field_key, concat_col in concat_columns.items():
                hash_col = f"__hash_{field_key}"
                # md5() in DuckDB returns hex string, cast to VARCHAR for consistency
                hash_expr = f"CAST(md5({concat_col}) AS VARCHAR)"
                hash_selects.append(f"{hash_expr} as {hash_col}")

            hash_clause = ", ".join(hash_selects)
            table_sql = table.compile()
            return f"SELECT *, {hash_clause} FROM ({table_sql}) AS __metaxy_temp"

        generators[HashAlgorithm.MD5] = md5_generator

        # Check if hashfuncs extension is in the list
        extension_names = []
        for ext in self.extensions:
            if isinstance(ext, str):
                extension_names.append(ext)
            elif isinstance(ext, Mapping):
                name = ext.get("name")
                if not name:
                    raise ValueError(
                        f"Extension mapping must have a non-empty 'name' key; got {ext}"
                    )
                extension_names.append(str(name))
            else:
                # Must be ExtensionSpec
                from metaxy.metadata_store.duckdb import ExtensionSpec

                if isinstance(ext, ExtensionSpec):
                    extension_names.append(ext.name)
                else:
                    raise TypeError(
                        f"Extension must be str, Mapping, or ExtensionSpec; got {type(ext)}"
                    )

        if "hashfuncs" in extension_names:

            def xxhash32_generator(table, concat_columns: dict[str, str]) -> str:
                hash_selects: list[str] = []
                for field_key, concat_col in concat_columns.items():
                    hash_col = f"__hash_{field_key}"
                    hash_expr = f"CAST(xxh32({concat_col}) AS VARCHAR)"
                    hash_selects.append(f"{hash_expr} as {hash_col}")

                hash_clause = ", ".join(hash_selects)
                table_sql = table.compile()
                return f"SELECT *, {hash_clause} FROM ({table_sql}) AS __metaxy_temp"

            def xxhash64_generator(table, concat_columns: dict[str, str]) -> str:
                hash_selects: list[str] = []
                for field_key, concat_col in concat_columns.items():
                    hash_col = f"__hash_{field_key}"
                    hash_expr = f"CAST(xxh64({concat_col}) AS VARCHAR)"
                    hash_selects.append(f"{hash_expr} as {hash_col}")

                hash_clause = ", ".join(hash_selects)
                table_sql = table.compile()
                return f"SELECT *, {hash_clause} FROM ({table_sql}) AS __metaxy_temp"

            generators[HashAlgorithm.XXHASH32] = xxhash32_generator
            generators[HashAlgorithm.XXHASH64] = xxhash64_generator

        return generators

    @property
    def supported_algorithms(self) -> list[HashAlgorithm]:
        """Algorithms supported by this calculator based on loaded extensions."""
        # Dynamically determine based on what was actually loaded
        return list(self._hash_sql_generators.keys())

    @property
    def default_algorithm(self) -> HashAlgorithm:
        """Default hash algorithm for DuckDB.

        Uses XXHASH64 if hashfuncs extension is loaded, otherwise MD5.
        """
        if HashAlgorithm.XXHASH64 in self.supported_algorithms:
            return HashAlgorithm.XXHASH64
        return HashAlgorithm.MD5
