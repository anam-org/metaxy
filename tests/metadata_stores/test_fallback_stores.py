"""Test fallback store behavior and warnings.

Tests that resolve_update correctly switches between native and Polars components
based on whether upstream features are in fallback stores, and that appropriate
warnings are issued.
"""

from typing import Any

import narwhals as nw
import polars as pl
import pytest
from pytest_cases import parametrize_with_cases

from metaxy import (
    Feature,
    FeatureDep,
    FeatureKey,
    FeatureSpec,
    FieldDep,
    FieldKey,
    FieldSpec,
)
from metaxy._testing import HashAlgorithmCases, assert_all_results_equal
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.metadata_store import (
    HashAlgorithmNotSupportedError,
    InMemoryMetadataStore,
    MetadataStore,
)
from metaxy.metadata_store.duckdb import DuckDBMetadataStore
from metaxy.models.feature import FeatureGraph

# ============= HELPERS =============


def get_available_store_types_for_fallback() -> list[str]:
    """Get store types that support native components and can be used for fallback tests.

    We test with DuckDB since it:
    1. Supports native components (IbisDataVersionCalculator)
    2. Can be easily instantiated with separate databases
    3. Supports multiple hash algorithms
    """
    return ["duckdb"]


def create_store_for_fallback(
    store_type: str,
    prefer_native: bool,
    hash_algorithm: HashAlgorithm,
    params: dict[str, Any],
    suffix: str = "",
    fallback_stores: list[MetadataStore] | None = None,
) -> MetadataStore:
    """Create a store instance for fallback testing.

    Args:
        store_type: "duckdb"
        prefer_native: Whether to prefer native components
        hash_algorithm: Hash algorithm to use
        params: Store-specific parameters
        suffix: Suffix to add to database filename (for creating distinct stores)
        fallback_stores: Optional list of fallback stores to configure
    """
    tmp_path = params.get("tmp_path")
    assert tmp_path is not None, f"tmp_path parameter required for {store_type}"

    if store_type == "duckdb":
        db_path = tmp_path / f"fallback_test_{suffix}_{hash_algorithm.value}.duckdb"
        extensions: list[str] = (
            ["hashfuncs"]
            if hash_algorithm in [HashAlgorithm.XXHASH32, HashAlgorithm.XXHASH64]
            else []
        )
        return DuckDBMetadataStore(
            db_path,
            hash_algorithm=hash_algorithm,
            extensions=extensions,  # pyright: ignore[reportArgumentType]
            prefer_native=prefer_native,
            fallback_stores=fallback_stores,
        )
    else:
        raise ValueError(f"Unknown store type for fallback test: {store_type}")


# ============= FEATURE DEFINITIONS =============


class RootFeature(
    Feature,
    spec=FeatureSpec(
        key=FeatureKey(["fallback_test", "root"]),
        deps=None,
        fields=[
            FieldSpec(key=FieldKey(["default"]), code_version=1),
        ],
    ),
):
    """Root feature with no dependencies."""

    pass


class DownstreamFeature(
    Feature,
    spec=FeatureSpec(
        key=FeatureKey(["fallback_test", "downstream"]),
        deps=[FeatureDep(key=FeatureKey(["fallback_test", "root"]))],
        fields=[
            FieldSpec(
                key=FieldKey(["default"]),
                code_version=1,
                deps=[
                    FieldDep(
                        feature_key=FeatureKey(["fallback_test", "root"]),
                        fields=[FieldKey(["default"])],
                    )
                ],
            ),
        ],
    ),
):
    """Downstream feature depending on RootFeature."""

    pass


# ============= TESTS =============


@parametrize_with_cases("hash_algorithm", cases=HashAlgorithmCases)
@pytest.mark.parametrize("primary_store_type", get_available_store_types_for_fallback())
@pytest.mark.parametrize(
    "fallback_store_type", ["inmemory"]
)  # Parametrized for future expansion
def test_fallback_store_warning_issued(
    store_params: dict[str, Any],
    hash_algorithm: HashAlgorithm,
    primary_store_type: str,
    fallback_store_type: str,
    caplog,
    snapshot,
):
    """Test that warning IS issued when upstream feature is in fallback store.

    This tests the core fallback scenario:
    - Root feature is in fallback_store (InMemory - no native components)
    - Downstream feature will be written to primary_store (DuckDB - has native)
    - resolve_update() should switch to Polars components and issue warning
    - Results should match snapshot across different configurations
    """
    graph = FeatureGraph()
    graph.add_feature(RootFeature)
    graph.add_feature(DownstreamFeature)

    # Create fallback store (InMemory - simple, no native components)
    fallback_store = InMemoryMetadataStore(hash_algorithm=hash_algorithm)

    # Prepare root data
    root_data = pl.DataFrame(
        {
            "sample_uid": [1, 2, 3],
            "data_version": [
                {"default": "hash1"},
                {"default": "hash2"},
                {"default": "hash3"},
            ],
        }
    )

    results = {}

    try:
        # Setup: Write root feature to fallback store
        with fallback_store, graph.use():
            fallback_store.write_metadata(RootFeature, root_data)

        # Test with prefer_native=True and prefer_native=False
        for prefer_native in [True, False]:
            # Create primary store with native component support and fallback configured
            primary_store = create_store_for_fallback(
                primary_store_type,
                prefer_native=prefer_native,
                hash_algorithm=hash_algorithm,
                params=store_params,
                suffix=f"primary_native{prefer_native}",
                fallback_stores=[fallback_store],
            )

            # Test: Resolve downstream feature with primary store that has fallback
            with primary_store, fallback_store, graph.use():
                import logging

                caplog.clear()
                with caplog.at_level(logging.WARNING):
                    result = primary_store.resolve_update(DownstreamFeature)

                # Verify warning was issued only when prefer_native=True
                fallback_warnings = [
                    record
                    for record in caplog.records
                    if "upstream dependencies in fallback stores" in record.message
                    and "Falling back to in-memory Polars processing" in record.message
                ]

                if prefer_native:
                    assert len(fallback_warnings) == 1, (
                        f"Expected exactly 1 fallback warning for {primary_store_type} with prefer_native=True, "
                        f"but got {len(fallback_warnings)}: {[r.message for r in fallback_warnings]}"
                    )
                else:
                    # prefer_native=False means we intentionally use Polars, not a fallback
                    assert len(fallback_warnings) == 0, (
                        f"Unexpected fallback warning for {primary_store_type} with prefer_native=False"
                    )

                # Collect data versions for comparison
                added_sorted = (
                    result.added.to_polars()
                    if isinstance(result.added, nw.DataFrame)
                    else result.added
                ).sort("sample_uid")
                versions = added_sorted["data_version"].to_list()

                results[(primary_store_type, fallback_store_type, prefer_native)] = {
                    "added": len(result.added),
                    "changed": len(result.changed),
                    "removed": len(result.removed),
                    "versions": versions,
                }

    except HashAlgorithmNotSupportedError:
        pytest.skip(f"Hash algorithm {hash_algorithm} not supported")

    # All variants should produce identical results (including data versions)
    assert_all_results_equal(results, snapshot)

    # Verify expected behavior
    if results:
        reference = list(results.values())[0]
        assert reference["added"] == 3
        assert reference["changed"] == 0
        assert reference["removed"] == 0
        assert len(reference["versions"]) == 3
        assert all(isinstance(v, dict) for v in reference["versions"])


@parametrize_with_cases("hash_algorithm", cases=HashAlgorithmCases)
@pytest.mark.parametrize("store_type", get_available_store_types_for_fallback())
def test_no_fallback_warning_when_all_local(
    store_params: dict[str, Any],
    hash_algorithm: HashAlgorithm,
    store_type: str,
    caplog,
):
    """Test that warning is NOT issued when all upstream is in the same store.

    This tests the normal case:
    - Both root and downstream features are in the same store
    - resolve_update() should use native components (no Polars fallback)
    - No warning should be issued
    """
    graph = FeatureGraph()
    graph.add_feature(RootFeature)
    graph.add_feature(DownstreamFeature)

    store = create_store_for_fallback(
        store_type,
        prefer_native=True,
        hash_algorithm=hash_algorithm,
        params=store_params,
        suffix="single",
    )

    try:
        with store, graph.use():
            # Write root feature to same store
            root_data = pl.DataFrame(
                {
                    "sample_uid": [1, 2, 3],
                    "data_version": [
                        {"default": "hash1"},
                        {"default": "hash2"},
                        {"default": "hash3"},
                    ],
                }
            )
            store.write_metadata(RootFeature, root_data)

            # Resolve downstream feature - all upstream is local
            import logging

            caplog.clear()
            with caplog.at_level(logging.WARNING):
                result = store.resolve_update(DownstreamFeature)

            # Verify NO fallback warning was issued
            fallback_warnings = [
                record
                for record in caplog.records
                if "upstream dependencies in fallback stores" in record.message
                or "Falling back to in-memory Polars processing" in record.message
            ]

            assert len(fallback_warnings) == 0, (
                f"Unexpected fallback warning for {store_type} when all upstream is local: "
                f"{[r.message for r in fallback_warnings]}"
            )

            # Verify results are still correct
            assert len(result.added) == 3
            assert len(result.changed) == 0
            assert len(result.removed) == 0

    except HashAlgorithmNotSupportedError:
        pytest.skip(f"Hash algorithm {hash_algorithm} not supported")


@parametrize_with_cases("hash_algorithm", cases=HashAlgorithmCases)
@pytest.mark.parametrize("primary_store_type", get_available_store_types_for_fallback())
@pytest.mark.parametrize("fallback_store_type", ["inmemory"])
def test_fallback_store_switches_to_polars_components(
    store_params: dict[str, Any],
    hash_algorithm: HashAlgorithm,
    primary_store_type: str,
    fallback_store_type: str,
    caplog,
    snapshot,
):
    """Test that resolve_update switches from native to Polars components with fallback.

    This verifies the component selection logic:
    1. When all upstream is local → use native components (no warning)
    2. When upstream in fallback → switch to Polars components (with warning)
    3. Results should be identical in both cases
    4. All scenarios are snapshotted for regression detection
    """
    graph = FeatureGraph()
    graph.add_feature(RootFeature)
    graph.add_feature(DownstreamFeature)

    # Setup root feature data
    root_data = pl.DataFrame(
        {
            "sample_uid": [1, 2, 3],
            "data_version": [
                {"default": "hash1"},
                {"default": "hash2"},
                {"default": "hash3"},
            ],
        }
    )

    results: dict[Any, Any] = {}

    # Scenario 1: All local (native components)
    store_all_local = create_store_for_fallback(
        primary_store_type,
        prefer_native=True,
        hash_algorithm=hash_algorithm,
        params=store_params,
        suffix="all_local",
    )

    try:
        with store_all_local, graph.use():
            store_all_local.write_metadata(RootFeature, root_data)

            import logging

            caplog.clear()
            with caplog.at_level(logging.WARNING):
                result_local = store_all_local.resolve_update(DownstreamFeature)

            # Should be no warnings
            warnings_local = [
                r for r in caplog.records if "fallback" in r.message.lower()
            ]
            assert len(warnings_local) == 0

            # Collect data versions from result
            added_local = (
                result_local.added.to_polars()
                if isinstance(result_local.added, nw.DataFrame)
                else result_local.added
            )
            versions_local = added_local.sort("sample_uid")["data_version"].to_list()

            results[(primary_store_type, fallback_store_type, "all_local")] = {
                "added": len(result_local.added),
                "changed": len(result_local.changed),
                "removed": len(result_local.removed),
                "versions": versions_local,
            }

    except HashAlgorithmNotSupportedError:
        pytest.skip(f"Hash algorithm {hash_algorithm} not supported")
        return

    # Scenario 2: Upstream in fallback (Polars components)
    # Use InMemory for fallback (realistic - simple store from previous deployment)
    fallback_store = InMemoryMetadataStore(hash_algorithm=hash_algorithm)

    # Write root to fallback store
    with fallback_store, graph.use():
        fallback_store.write_metadata(RootFeature, root_data)

    # Create primary store with fallback configured
    primary_store = create_store_for_fallback(
        primary_store_type,
        prefer_native=True,
        hash_algorithm=hash_algorithm,
        params=store_params,
        suffix="with_fallback",
        fallback_stores=[fallback_store],
    )

    with primary_store, fallback_store, graph.use():
        import logging

        caplog.clear()
        with caplog.at_level(logging.WARNING):
            result_fallback = primary_store.resolve_update(DownstreamFeature)

        # Should have warning about fallback
        warnings_fallback = [
            r
            for r in caplog.records
            if "upstream dependencies in fallback stores" in r.message
        ]
        assert len(warnings_fallback) == 1

        # Collect data versions from result
        added_fallback = (
            result_fallback.added.to_polars()
            if isinstance(result_fallback.added, nw.DataFrame)
            else result_fallback.added
        )
        versions_fallback = added_fallback.sort("sample_uid")["data_version"].to_list()

        results[(primary_store_type, fallback_store_type, "with_fallback")] = {
            "added": len(result_fallback.added),
            "changed": len(result_fallback.changed),
            "removed": len(result_fallback.removed),
            "versions": versions_fallback,
        }

    # Compare results - versions should be identical
    assert_all_results_equal(results, snapshot)

    # Verify both scenarios have correct results
    assert len(results) == 2
    for scenario_key, result in results.items():
        result_versions = result["versions"]
        assert result["added"] == 3
        assert result["changed"] == 0
        assert result["removed"] == 0
        assert len(result_versions) == 3
        assert all(isinstance(v, dict) for v in result_versions)


@parametrize_with_cases("hash_algorithm", cases=HashAlgorithmCases)
@pytest.mark.parametrize("store_type", get_available_store_types_for_fallback())
def test_prefer_native_false_no_warning_even_without_fallback(
    store_params: dict[str, Any],
    hash_algorithm: HashAlgorithm,
    store_type: str,
    caplog,
):
    """Test that prefer_native=False doesn't issue fallback warning.

    When prefer_native=False, the store always uses Polars components,
    so there's no "fallback" from native to Polars - it's intentional.
    The warning should only be issued when we CAN'T use native due to fallback stores.
    """
    graph = FeatureGraph()
    graph.add_feature(RootFeature)
    graph.add_feature(DownstreamFeature)

    # Create store with prefer_native=False
    store = create_store_for_fallback(
        store_type,
        prefer_native=False,  # Explicitly disable native
        hash_algorithm=hash_algorithm,
        params=store_params,
        suffix="no_native",
    )

    try:
        with store, graph.use():
            root_data = pl.DataFrame(
                {
                    "sample_uid": [1, 2, 3],
                    "data_version": [
                        {"default": "hash1"},
                        {"default": "hash2"},
                        {"default": "hash3"},
                    ],
                }
            )
            store.write_metadata(RootFeature, root_data)

            import logging

            caplog.clear()
            with caplog.at_level(logging.WARNING):
                result = store.resolve_update(DownstreamFeature)

            # Should be no warnings - prefer_native=False is intentional, not a fallback
            warnings = [r for r in caplog.records]
            assert len(warnings) == 0, (
                f"Unexpected warnings with prefer_native=False: {[r.message for r in warnings]}"
            )

            # Verify results are correct
            assert len(result.added) == 3

    except HashAlgorithmNotSupportedError:
        pytest.skip(f"Hash algorithm {hash_algorithm} not supported")
