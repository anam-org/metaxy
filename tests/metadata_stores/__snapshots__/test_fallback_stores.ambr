# serializer version: 1
# name: test_fallback_store_switches_to_polars_components[inmemory-duckdb-md5]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      'all_local',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': 'cfc410b7d4428aeb47c3e69793831e1d',
        }),
        dict({
          'default': '7a06ff64cf44a164b8c9032785420f9a',
        }),
        dict({
          'default': '1a2984d8cbb982cd7adbf302d1fbbfb4',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      'with_fallback',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': 'cfc410b7d4428aeb47c3e69793831e1d',
        }),
        dict({
          'default': '7a06ff64cf44a164b8c9032785420f9a',
        }),
        dict({
          'default': '1a2984d8cbb982cd7adbf302d1fbbfb4',
        }),
      ]),
    }),
  })
# ---
# name: test_fallback_store_switches_to_polars_components[inmemory-duckdb-xxhash32]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      'all_local',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '820994320',
        }),
        dict({
          'default': '1359843373',
        }),
        dict({
          'default': '3025128646',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      'with_fallback',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '820994320',
        }),
        dict({
          'default': '1359843373',
        }),
        dict({
          'default': '3025128646',
        }),
      ]),
    }),
  })
# ---
# name: test_fallback_store_switches_to_polars_components[inmemory-duckdb-xxhash64]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      'all_local',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '8781274474994251810',
        }),
        dict({
          'default': '2477696910468939981',
        }),
        dict({
          'default': '7707417068700355918',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      'with_fallback',
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '8781274474994251810',
        }),
        dict({
          'default': '2477696910468939981',
        }),
        dict({
          'default': '7707417068700355918',
        }),
      ]),
    }),
  })
# ---
# name: test_fallback_store_warning_issued[inmemory-duckdb-md5]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      False,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': 'cfc410b7d4428aeb47c3e69793831e1d',
        }),
        dict({
          'default': '7a06ff64cf44a164b8c9032785420f9a',
        }),
        dict({
          'default': '1a2984d8cbb982cd7adbf302d1fbbfb4',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      True,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': 'cfc410b7d4428aeb47c3e69793831e1d',
        }),
        dict({
          'default': '7a06ff64cf44a164b8c9032785420f9a',
        }),
        dict({
          'default': '1a2984d8cbb982cd7adbf302d1fbbfb4',
        }),
      ]),
    }),
  })
# ---
# name: test_fallback_store_warning_issued[inmemory-duckdb-xxhash32]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      False,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '820994320',
        }),
        dict({
          'default': '1359843373',
        }),
        dict({
          'default': '3025128646',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      True,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '820994320',
        }),
        dict({
          'default': '1359843373',
        }),
        dict({
          'default': '3025128646',
        }),
      ]),
    }),
  })
# ---
# name: test_fallback_store_warning_issued[inmemory-duckdb-xxhash64]
  dict({
    tuple(
      'duckdb',
      'inmemory',
      False,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '8781274474994251810',
        }),
        dict({
          'default': '2477696910468939981',
        }),
        dict({
          'default': '7707417068700355918',
        }),
      ]),
    }),
    tuple(
      'duckdb',
      'inmemory',
      True,
    ): dict({
      'added': 3,
      'changed': 0,
      'removed': 0,
      'versions': list([
        dict({
          'default': '8781274474994251810',
        }),
        dict({
          'default': '2477696910468939981',
        }),
        dict({
          'default': '7707417068700355918',
        }),
      ]),
    }),
  })
# ---
